oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,210 @@
+
+  | ``header.search``
+  | ``login.no.such.password``
+  | ``login.privacy.policy``
+
+The part of the id used for namespacing is purely to reduce collisions and help human readability; it's not intended to be programmatically parsable. That said, a helpful convention is to make it the name of your application or repo. Note that if you end up with a duplicate message id in the same translation project in Transifex, and they have different content, one of them will "win" when you download your translated strings, leading to an out-of-place string in your application.
+
+Alternatives Considered
+-----------------------
+
+Previous i18n implementations have sometimes used the English text as the message id. While this saves developers a step in their work, it has the drawback that a change in the English copy invalidates all existing translations, even if they would not otherwise have needed to change. (For example, capitalizing a noun in English has no impact on the German translation, because nouns are capitalized in German anyway. Or an English change might be correcting a typo that was correct in the translations.) Do not use English copy as the message id.
+
+Cross-Site Scripting Vulnerability
+==================================
+
+Because translated strings often contain placeholders that will be filled in at render time, possibly using user input, localization is a potential opening for cross-site scripting (XSS) attacks. The react-intl functions exposed by frontend-i18n are be XSS-safe.
+
+Mapping Server Codes To Messages
+================================
+
+When a server response can contain one of a set of enumerated values, it can be tempting to assemble your message id on the fly by concatenating a common prefix and the server-provided value:
+
+``const id = 'myapp.errors.${errorCode}';``
+
+However, an unexpected value from the server can cause this technique to generate a message id that doesn't exist. If you choose to do this, it's best to whitelist your server-provided value against a list of expected responses.
+
+Missing Translations
+====================
+
+Missing translations can be a disorienting experience for a user. We strongly suggest a careful evaluation of translation coverage, or a clear indicator of beta status, before making a new language available to the public.
+
+****************
+Server-Side i18n
+****************
+
+Have your server send symbolic values or error codes, rather than translated display text, to convey information, and then localize on the front end. This keeps all of your localization together, simplifies testing your server code, and makes your server response reusable by other front ends, which may wish to display the same information but with different text.
```

Comment:
Is this section really just mean we should localize everything on the backend server, and have no logic on the javascript side to do translation? If my understanding is correct, I feel the existing way to phrase is confusing.

Reply:
@schenedx No, the opposite.  The server should provide programmatic information only, and leave localization to the front end.

Reply:
I see. I think a graphical representation of what this mechanism is may bring more clarity. Like draw two boxes, one the backend server, one the frontend. Then have the non-localized data transfered to front end. And have the localized data shown to users.